File: core/src/main/java/y2k/joyreactor/images/MultiTryDownloader.java
package y2k.joyreactor.images;

import rx.Observable;
import rx.Subscriber;
import y2k.joyreactor.ForegroundScheduler;
import y2k.joyreactor.http.HttpClient;

import java.io.*;
import java.util.concurrent.TimeUnit;

/**
 * Created by y2k on 12/10/15.
 */
class MultiTryDownloader {

    private static TaskExecutor DOWNLOAD_EXECUTOR = new TaskExecutor(3);
    private static final int MAX_TRY = 5;

    private File dir;
    private String url;

    MultiTryDownloader(File dir, String url) {
        this.dir = dir;
        this.url = url;
    }

    public Observable<File> downloadAsync() {
        return Observable.create(subscriber -> downloadAsync(0, subscriber));
    }

    private void downloadAsync(int tryNumber, Subscriber<? super File> subscriber) {
        ForegroundScheduler.getInstance().createWorker().schedule(() -> {
            DOWNLOAD_EXECUTOR.execute(() -> {
                try {
                    subscriber.onNext(downloadToTempFile());
                    subscriber.onCompleted();
                } catch (Exception e) {
                    if (tryNumber > MAX_TRY) subscriber.onError(e);
                    else downloadAsync(tryNumber + 1, subscriber);
                }
            });
        }, 250 << tryNumber, TimeUnit.MILLISECONDS);
    }

    private File downloadToTempFile() throws IOException {
        File result = null;
        try {
            result = File.createTempFile("download_", null, dir);
            HttpClient.getInstance().downloadToFile(url, result);
            return result;
        } catch (IOException e) {
            if (result != null) result.delete();
            throw e;
        }
    }
}

File: core/src/main/kotlin/y2k/joyreactor/images/MultiTryDownloader.kt
package y2k.joyreactor.images

import rx.Single
import rx.SingleSubscriber
import y2k.joyreactor.common.schedule
import y2k.joyreactor.http.HttpClient
import java.io.File
import java.io.IOException
import java.util.concurrent.Executors
import java.util.concurrent.LinkedBlockingDeque
import java.util.concurrent.ThreadPoolExecutor
import java.util.concurrent.TimeUnit

/**
 * Created by y2k on 12/10/15.
 */
internal class MultiTryDownloader(
    private val httpClient: HttpClient,
    private val dir: File,
    private val url: String) {

    fun downloadAsync(): Single<File> {
        return Single.create<File> { downloadAsync(0, it) }
    }

    private fun downloadAsync(tryNumber: Int, subscriber: SingleSubscriber<in File>) {
        TIMER.schedule(250L shl tryNumber) {
            DOWNLOAD_EXECUTOR.execute {
                try {
                    subscriber.onSuccess(downloadToTempFile())
                } catch (e: Exception) {
                    if (tryNumber >= MAX_RETRY) subscriber.onError(e)
                    else downloadAsync(tryNumber + 1, subscriber)
                }
            }
        }
    }

    private fun downloadToTempFile(): File {
        var result: File? = null
        try {
            result = File.createTempFile("download_", null, dir)
            httpClient.downloadToFile(url, result, null)
            return result
        } catch (e: IOException) {
            if (result != null) result.delete()
            throw e
        }
    }

    companion object {

        private val MAX_RETRY = 5
        private val MAX_THREADS = 5
        private val TIMER = Executors.newSingleThreadScheduledExecutor()
        private val DOWNLOAD_EXECUTOR = ThreadPoolExecutor(MAX_THREADS, MAX_THREADS, 1, TimeUnit.SECONDS,
            object : LinkedBlockingDeque<Runnable>() {

                override fun offer(e: Runnable?): Boolean {
                    return super.offerFirst(e)
                }
            }).apply { allowCoreThreadTimeOut(true) }
    }
}